python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -280,6 +284,62 @@ def _update(self):
         for node_id in nodes:
             self.recover_if_needed(node_id, now)
 
+    def _get_concurrent_resource_demand_to_launch(
+            self, to_launch: Dict[NodeType, int],
+            non_terminated_nodes: List[NodeID],
+            pending_launches_nodes: Dict[NodeType, int]
+    ) -> Dict[NodeType, int]:
+        """Updates the max concurrent resources to launch for each node type.
+
+        Given the current nodes that should be launched, the non terminated
+        nodes (running and pending) and the pending to be launched nodes. This
+        method calculates the maximum number of nodes to launch concurrently
+        for each node type as follows:
+            1) Calculates the running nodes.
+            2) Calculates the pending nodes and gets the launching nodes.
+            3) Limits the total number of pending + currently-launching +
+               to-be-launched nodes to max(5, frac * running_nodes[node_type]).
+
+        Args:
+            to_launch: Number of nodes to launch based on resource demand.
+            non_terminated_nodes: Non terminated nodes (pending/running).
+            pending_launches_nodes: Nodes that are in the launch queue.
+        Returns:
+            Dict[NodeType, int]: Maximum number of nodes to launch for each
+                node type.
+        """
+        running_nodes = collections.defaultdict(int)
+        pending_nodes = collections.defaultdict(int)
+        for node_id in non_terminated_nodes:
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_USER_NODE_TYPE in tags:
+                node_type = tags[TAG_RAY_USER_NODE_TYPE]
+                if self.provider.is_running(node_id):
```

Comment:
Is this really enough? A running node isn't necessarily part of the cluster yet (because it could still be installing dependencies).

Reply:
What else do you suggest?

Reply:
Probably better to check if `TAG_RAY_NODE_STATUS` is `STATUS_UP_TO_DATE`.

Reply:
The conclusion here is that if it is not STATUS_UP_TO_DATE then it is pending? This will make it even more slower to start instances.

Reply:
This isn't about speed, it's about correctness. If the status isn't up to date, the node isn't in the cluster yet. It's just a random EC2 instance that isn't doing anything useful.

Reply:
Can you check the updated one? @wuisawesome